domain/model/service/migration.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/clock"
+	"github.com/juju/version/v2"
+
+	coreconstraints "github.com/juju/juju/core/constraints"
+	coremodel "github.com/juju/juju/core/model"
+	"github.com/juju/juju/domain/constraints"
+	modelerrors "github.com/juju/juju/domain/model/errors"
+	"github.com/juju/juju/internal/errors"
+	"github.com/juju/juju/internal/uuid"
+)
+
+// MigrationService defines a service for model migration operations.
+type MigrationService struct {
+	clock                 clock.Clock
+	modelID               coremodel.UUID
+	controllerSt          ControllerState
+	modelSt               ModelState
+	environProviderGetter EnvironVersionProviderFunc
+	agentBinaryFinder     AgentBinaryFinder
+}
+
+// NewMigrationService creates a new instance of MigrationService.
+func NewMigrationService(
```

Comment:
@ycliuhw can we please move this to a separate PR? It deserves to be reviewed independently of this change.

Reply:
It has to be done in this PR because https://github.com/juju/juju/pull/19012#discussion_r1974762842

Reply:
Ahh right. Well then it isn't a drive-by and is required.

Reply:
Ah, right, rephrased the description.